Update cloudscale.ch terraform module to `v3.1.1` (#25)
appuio/component-openshift4-terraform: docs/modules/ROOT/pages/how-tos/upgrade-cloudscale-v2-v3.adoc

= Upgrade cloudscale.ch Terraform module from v2 to v3

> https://github.com/appuio/terraform-openshift4-cloudscale/releases/tag/v3.0.0[terraform-openshift4-cloudscale v3.0.0] moves load balancer state to a shared module.

Reviewer: ```suggestion
When migrating from v2 to v3, the following breaking changes require manual changes to the Terraform state of existing clusters:

* https://github.com/appuio/terraform-openshift4-cloudscale/releases/tag/v3.0.0[terraform-openshift4-cloudscale v3.0.0] uses the new https://github.com/appuio/terraform-modules/tree/main/modules/vshn-lbaas-cloudscale[shared load-balancer module] to provision the LBs.
* https://github.com/appuio/terraform-openshift4-cloudscale/releases/tag/v3.1.0[terraform-openshift4-cloudscale v3.1.0] adds support for using an existing cloudscale.ch subnet.
```

> That release includes breaking changes which require manual steps to upgrade existing clusters.

Reviewer: ```suggestion
```

export GITLAB_REPOSITORY_NAME=${GITLAB_REPOSITORY_URL##*/}
export GITLAB_CATALOG_PROJECT_ID=$(curl -sH "Authorization: Bearer ${GITLAB_TOKEN}" "https://git.vshn.net/api/v4/projects?simple=true&search=${GITLAB_REPOSITORY_NAME/.git}" | jq -r ".[] | select(.ssh_url_to_repo == \"${GITLAB_REPOSITORY_URL}\") | .id")
export GITLAB_STATE_URL="https://git.vshn.net/api/v4/projects/${GITLAB_CATALOG_PROJECT_ID}/terraform/state/cluster"

pushd catalog/manifests/openshift4-terraform/
----
+
.Initialize Terraform
[source,bash]
----
terraform init \
  "-backend-config=address=${GITLAB_STATE_URL}" \
  "-backend-config=lock_address=${GITLAB_STATE_URL}/lock" \
  "-backend-config=unlock_address=${GITLAB_STATE_URL}/lock" \
  "-backend-config=username=${GITLAB_USER}" \
  "-backend-config=password=${GITLAB_TOKEN}" \
  "-backend-config=lock_method=POST" \
  "-backend-config=unlock_method=DELETE" \
  "-backend-config=retry_wait_min=5"
----

. Migrate state
+
.Run the migration script
[source,bash]
----
./migrate-state-v2-v3.sh
----
+
.Verify state using `plan`
[source,bash]
----
# No resources, except the hieradata git checkout, should be recreated.
terraform plan
----

> NOTE: Terraform may wants to create the hieradata git checkout. This is expected and can be ignored.

Reviewer: ```suggestion
NOTE: Terraform may want to create the hieradata git checkout. This is expected and can be ignored.
```
